L1Trigger/L1THGCal/src/HGCalTriggerTowerGeometryHelper.cc

```diff
@@ -117,21 +121,89 @@ unsigned short HGCalTriggerTowerGeometryHelper::getTriggerTowerFromEtaPhi(const
   unsigned int trigger_cell_id = thecell.detId();
   // NOTE: if the TC is not found in the map than it is mapped via eta-phi coords.
   // this can be considered dangerous (silent failure of the map) but it actually allows to save
   // memory mapping explicitly only what is actually needed
   auto tower_id_itr = cells_to_trigger_towers_.find(trigger_cell_id);
-  if (tower_id_itr != cells_to_trigger_towers_.end())
-    return tower_id_itr->second;
-
-  return getTriggerTowerFromEtaPhi(thecell.position().eta(), thecell.position().phi());
+  if (tower_id_itr != cells_to_trigger_towers_.end()){
+    binIDandShares.insert({tower_id_itr->second, 1.0});
+    return binIDandShares;
+  }
+  binIDandShares.insert({getTriggerTowerFromEtaPhi(thecell.position().eta(), thecell.position().phi()), 1.0});
+  return binIDandShares;
 }
 
-unsigned short HGCalTriggerTowerGeometryHelper::getTriggerTower(const l1t::HGCalTriggerSums& thesum) const {
-  return getTriggerTowerFromEtaPhi(thesum.position().eta(), thesum.position().phi());
+std::unordered_map<unsigned short, float> HGCalTriggerTowerGeometryHelper::getTriggerTower(const l1t::HGCalTriggerSums& thesum) const {
+  std::unordered_map<unsigned short, float> binIDandShares = {}; 
+  if(!splitModuleSum_){
+    binIDandShares.insert({getTriggerTowerFromEtaPhi(thesum.position().eta(), thesum.position().phi()), 1.0});
+    return binIDandShares;
+  } else{
+    HGCalModuleDetId detid(thesum.detId());
+    int moduleU = detid.moduleU();
+    int moduleV = detid.moduleV();
+    int layer = detid.layer();
+    int sector = detid.sector();
+    int zside = detid.zside();
+
+    int subdet = -1;
+    int splitDivisor = -1;
+    if(detid.isHScintillator()){ //FIXME HFNose not handled
+      subdet = 2;
+      layer += 28;
+      splitDivisor = splitDivisorScint_; 
+    } else if(detid.isEE()){
+      subdet = 0;
+      splitDivisor = splitDivisorSilic_;
+    } else if(detid.isHSilicon()) {
+      subdet = 1;
+      layer += 28;
+      splitDivisor = splitDivisorSilic_;
+    }
+    
+    std::ifstream moduleTowerMappingStream(moduleTowerMapping_.fullPath());
+    if (!moduleTowerMappingStream.is_open()) {
+      throw cms::Exception("MissingDataFile") << "Cannot open HGCalTowerMapProducer moduleTowerMapping file\n";
+    }
+
+    std::vector<std::string> result; // the line including towers and module sum shares for this module
+    std::string line;
+    getline(moduleTowerMappingStream, line); //To skip first row
+    for( std::string line; getline(moduleTowerMappingStream, line ); ){
+      std::stringstream ss(line);
+      while(ss.good()){
+        std::string substr;
+        std::getline(ss, substr, ' ');
+        result.push_back(substr);
+      }
+      if ((std::stoi(result.at(0))==subdet) && (std::stoi(result.at(1))==layer) && (std::stoi(result.at(2))==moduleU) && (std::stoi(result.at(3))==moduleV)){
+        break;
+      } else{
+        result.clear();
+      }
+    }
```

Comment:
All the mapping file reading should be included in the constructor. It cannot be redone each time `getTriggerTower` is called